image/generate.go

```diff
@@ -0,0 +1,362 @@
+	err := os.MkdirAll(filepath.Join(il.rootDir, "blobs", "sha256"), 0700)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(filepath.Join(il.rootDir, "refs"), 0700)
+	if err != nil {
+		return err
+	}
+
+	// create image layout file
+	err = createLayoutFile(il.rootDir)
+	if err != nil {
+		return err
+	}
+
+	// create image layer blob file.
+	desc, err := createImageLayerFile(il.rootDir, il.tarList)
+	if err != nil {
+		return err
+	}
+	il.manifest = strings.Replace(il.manifest, "<layer_digest>", desc.Digest, 1)
+	il.manifest = strings.Replace(il.manifest, "<layer_size>", strconv.FormatInt(desc.Size, 10), 1)
+
+	desc, err = createConfigFile(il.rootDir, il.config)
+	if err != nil {
+		return err
+	}
+	il.manifest = strings.Replace(il.manifest, "<config_digest>", desc.Digest, 1)
+	il.manifest = strings.Replace(il.manifest, "<config_size>", strconv.FormatInt(desc.Size, 10), 1)
+
+	// create manifest blob file
+	desc, err = createManifestFile(il.rootDir, il.manifest)
+	if err != nil {
+		return err
+	}
+
+	return createRefFile(il.rootDir, il.ref, desc)
+}
+
+func createLayoutFile(root string) error {
+	layoutPath := filepath.Join(root, "oci-layout")
+	f, err := os.Create(layoutPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(layoutStr)))
+	return err
+}
+
+func createRefFile(root, ref string, mft descriptor) error {
+	refpath := filepath.Join(root, "refs", ref)
+	f, err := os.Create(refpath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	refStr = strings.Replace(refStr, "<manifest_digest>", mft.Digest, -1)
+	refStr = strings.Replace(refStr, "<manifest_size>", strconv.FormatInt(mft.Size, 10), -1)
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(refStr)))
+	return err
+}
+
+func createManifestFile(root, str string) (descriptor, error) {
+	name := filepath.Join(root, "blobs", "sha256", "test-manifest")
+	f, err := os.Create(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(str)))
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	return createHashedBlob(name)
+}
+
+func createConfigFile(root, config string) (descriptor, error) {
+	name := filepath.Join(root, "blobs", "sha256", "test-config")
+	f, err := os.Create(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, bytes.NewBuffer([]byte(config)))
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	return createHashedBlob(name)
+}
+
+func createImageLayerFile(root string, list []tarContent) (descriptor, error) {
+	name := filepath.Join(root, "blobs", "sha256", "test-layer")
+	err := createTarBlob(name, list)
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	desc, err := createHashedBlob(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+
+	desc.MediaType = v1.MediaTypeImageLayer
+	return desc, nil
+}
+
+func createTarBlob(name string, list []tarContent) error {
```

Comment:
I'm holding out hope that #8 or something like that lands.  Until we get a public API for creating layers, I'd rather not drag `createTarBlob` out of `image_test.go`.